indexer/displacement_manager.hpp

```diff
@@ -0,0 +1,226 @@
+private:
+  CellFeaturePair m_pair;
+  uint32_t m_bucket;
+};
+static_assert(sizeof(CellFeatureBucketTuple) == 16, "");
+#ifndef OMIM_OS_LINUX
+static_assert(is_trivially_copyable<CellFeatureBucketTuple>::value, "");
+#endif
+
+/// Displacement manager filters incoming single-point features to simplify runtime
+/// feature visibility displacement.
+template <class TSorter>
+class DisplacementManager
+{
+public:
+  DisplacementManager(TSorter & sorter) : m_sorter(sorter) {}
+
+  /// Add feature at bucket (zoom) to displacable queue if possible. Pass to bucket otherwise.
+  template <class TFeature>
+  void Add(vector<int64_t> const & cells, uint32_t bucket, TFeature const & ft, uint32_t index)
+  {
+    if (bucket != scales::GetUpperScale() && IsDisplaceable(ft))
+    {
+      m_storage[bucket].emplace_back(cells, ft, index, bucket);
+      return;
+    }
+
+    for (auto const & cell : cells)
+      m_sorter.Add(CellFeatureBucketTuple(CellFeaturePair(cell, index), bucket));
+  }
+
+  /// Check features intersection and pass result to sorter.
+  void Displace()
+  {
+    m4::Tree<DisplaceableNode> acceptedNodes;
+    list<DisplaceableNode> deferredNodes;
+    for (uint32_t scale = 0; scale < scales::GetUpperScale(); ++scale)
+    {
+      // Initialize queue.
+      priority_queue<DisplaceableNode> displaceableQueue;
+      for (auto const & node : m_storage[scale])
+        displaceableQueue.push(node);
+      for (auto const & node : deferredNodes)
+        displaceableQueue.push(node);
+      deferredNodes.clear();
+
+      float const delta = CalculateDeltaForZoom(scale);
+      size_t accepted = 0;
+
+      // Process queue.
+      while (!displaceableQueue.empty())
+      {
+        DisplaceableNode const maxNode = displaceableQueue.top();
+        displaceableQueue.pop();
+
+        // Check if this node is displaced by higher features.
+        m2::RectD const displacementRect(maxNode.center, maxNode.center);
+        bool isDisplaced = false;
+        acceptedNodes.ForEachInRect(m2::Inflate(displacementRect, {delta, delta}),
+            [&isDisplaced, &maxNode, &delta, &scale](DisplaceableNode const & node)
+            {
+              if (maxNode.center.SquareLength(node.center) < delta * delta && node.maxScale > scale)
+                isDisplaced = true;
+            });
+        if (isDisplaced)
+        {
+          deferredNodes.push_back(maxNode);
+          continue;
+        }
+
+        // Add feature to index otherwise.
+        AddNodeToSorter(maxNode, scale);
+        acceptedNodes.Add(maxNode);
+        accepted++;
+      }
+
+      LOG(LINFO, ("Displacement for scale", scale, "Features accepted:", accepted,
+                  "Features discarded:", deferredNodes.size()));
+    }
+
+    // Add all fully displaced features to the bottom scale.
+    for (auto const & node : deferredNodes)
+      AddNodeToSorter(node, scales::GetUpperScale());
+  }
+
+private:
+  struct DisplaceableNode
+  {
+    uint32_t index;
```

Comment:
Ok. But this is struct members. Style guide affords us to use fields without m_ prefix.